### PR TITLE
planner: suppport left join in join reorder

### PR DESCRIPTION
**Pull request by winoros**
### What problem does this PR solve?

Issue Number: close https://github.com/pingcap/tidb/issues/9268 

Problem Summary:

### What is changed and how it works?

If the left join is not related with other left join. We can safely move it to the tail. And place the left joins by their size.

If the left join is related with other left join. It should be put after the related one.

What's Changed:

Add the above logic.

How it Works:

### Related changes

- Need to cherry-pick to the release branch

### Check List 

Tests 

- Unit test

Side effects

- Performance regression
    - Consumes more CPU

### Release note 

- Support join reorder for left outer join.

### CODE REVIEW

File: planner/core/rule_join_reorder.go [hunk 1]
```diff
@@ -89,6 +126,18 @@ func (s *joinReOrderSolver) optimizeRecursive(ctx sessionctx.Context, p LogicalP
 		if err != nil {
 			return nil, err
 		}
+		sort.Slice(leftJoinTabs, func(i, j int) bool {
+			if len(leftJoinTabs[j].eqCond) > 0 && leftJoinTabs[i].p.Schema().ColumnIndex(leftJoinTabs[j].eqCond[0].GetArgs()[0].(*expression.Column)) >= 0 {
```
Comment (eurekaka) [inline planner/core/rule_join_reorder.go:130]: Please add comments for these 2 heuristic rules.

File: planner/core/rule_join_reorder.go [hunk 2]
```diff
@@ -89,6 +126,18 @@ func (s *joinReOrderSolver) optimizeRecursive(ctx sessionctx.Context, p LogicalP
...
+				return true
+			}
+			if len(leftJoinTabs[i].eqCond) > 0 && len(leftJoinTabs[j].eqCond) == 0 {
+				return true
+			}
+			return false
```
Comment (eurekaka) [inline planner/core/rule_join_reorder.go:136]: Shall we compare the row count of these `leftJoinTabs` as well when deciding the order, i.e, if `leftJoinTabs[i].p.stats.RowCount` is smaller than `leftJoinTabs[j].p.stats.RowCount`, we put j after i?

**Review by eurekaka — COMMENTED** [review]
Can we support join reorder for queries containing right outer join as well?

**Comment by winoros** [comment]
@eurekaka Right join's case will make the join dependency a little complex. I implemented a local version and it failed some test cases. Still in progress

File: planner/core/rule_join_reorder.go (hunk 1 above)
Comment (winoros) [reply planner/core/rule_join_reorder.go:141]: added

**Review by eurekaka — COMMENTED** [review]
LGTM

File: go.mod [hunk 3]
```diff
@@ -62,6 +62,7 @@ require (
 	golang.org/x/tools v0.0.0-20200527183253-8e7acdbce89d
 	google.golang.org/grpc v1.26.0
 	gopkg.in/natefinch/lumberjack.v2 v2.0.0
+	honnef.co/go/tools v0.0.1-2020.1.5 // indirect
```
Comment (lzmhhh123) [inline go.mod:65]: Fix this.

**Review by lzmhhh123 — COMMENTED** [review]
Rest LGTM.

**Comment by zz-jason** [comment]
I'm going to close this PR since it hasn't been updated for a long time. feel free to reopen if you want to continue with it. thank you for your contribution.

**Comment by wankunde** [comment]
Hello, is there any updates for this?

**Pull request closed without merging**